### PR TITLE
Triangle: Return `null` in `getBarycoord()` if the triangle is degenerated. 

### PR DESCRIPTION
**Pull request by gkjohnson**
Related issue: https://github.com/gkjohnson/three-bvh-csg/issues/183

**Description**

Previously the getBarycoord function was providing a reasonable looking result where no reasonable value could be provided when a degenerate triangle was used. This makes tracking errors down extremely difficult. Now an error will be thrown if a degenerate triangle is used.

### CODE REVIEW

**Comment by Mugen87** [comment]
It seems the unit tests require an update now.

`Triangle.getBarycoord()` is also used in context of raycasting and I fear a less graceful approach could produce issues. If you think throwing an error is better to return and undefined result, callers of `Triangle.getBarycoord()` in the repository should implement an error handler. That is `Triangle.getInterpolation()` and `Triangle.containsPoint()`. I would not like to see unhandled exception when user perform raycasting, tbh.

**Comment by gkjohnson** [comment]
> If you think throwing an error is better to return and undefined result, callers of Triangle.getBarycoord() in the repository should implement an error handler.

I just changed the function to return null. But the reality of it is that all of these functions are already returning unreasonable results if they're not checking for degenerate triangles (none of them are). If we think this is a possibility then this these functions should already be handling these cases.

Documenting the possibility and returning null will hopefully make this easier to detect / understand. The biggest issue is that the result _looked_ reasonable.

File: src/math/Triangle.js [hunk 1]
```diff
@@ -60,9 +60,8 @@ class Triangle {
 		// collinear or singular triangle
 		if ( denom === 0 ) {
 
-			// arbitrary location outside of triangle?
-			// not sure if this is the best idea, maybe should be returning undefined
-			return target.set( - 2, - 1, - 1 );
+			target.set( 0, 0, 0 );
+			return null;
```
Comment (Mugen87) [inline src/math/Triangle.js:64]: @WestLangley Do you think this is okay? I don't know if `0, 0, 0` is better than `- 2, - 1, - 1` but I like the idea or returning `null` to indicate an erroneous edge case.

Comment (WestLangley) [reply src/math/Triangle.js:64]: How about
```js
target.set( 1, 0, 0 );

return null;
```

Comment (gkjohnson) [reply src/math/Triangle.js:64]: Can you explain you're reasoning? 1, 0, 0 is a reasonable result that could be returned by the function when using a valid triangle. I picked 0, 0, 0 because it's a value that cannot be produced in a valid scenario making it easier to see that a failure is occuring.

Comment (WestLangley) [reply src/math/Triangle.js:64]: You are returning `null` to indicate that the triangle is degenerate. The app can handle the returned result as it pleases.

Comment (Mugen87) [reply src/math/Triangle.js:64]: Against this backdrop I vote to keep the previous value. If a project specifically tested for `(- 2, - 1, - 1)` to detect an error, we do not break this usage.

Comment (WestLangley) [reply src/math/Triangle.js:64]: Hmm... Maybe `getInterpolation()` should also test for `null` in case the points are collinear?

I do not really like retuning nonsense values.

Comment (WestLangley) [reply src/math/Triangle.js:64]: >I'm just suggesting that if we're going to set the value to anything we should set [it] to 0, 0, 0

I was concerned that the other methods that call `getBarycoord()` will return nonsense if `getBarycoord()` returns nonsense.

```js
const t = new THREE.Triangle(); // degenerate
const p = new THREE.Vector3( 1, 1, 1 );

console.log( t.containsPoint( p ) ); // true
```

I think I'll support whatever you prefer to do here. :-)

Comment (Mugen87) [reply src/math/Triangle.js:64]: > Returning -2, -1, -1 was never a documented case.

Even if it is not documented, devs could look in the code see what's happening. But maybe I'm too overcautious here. We can try `(0,0,0)` and see if this causes any troubles.

**Review by Mugen87 — APPROVED** [review]

File: src/math/Triangle.js (hunk 1 above)
Comment (WestLangley) [reply src/math/Triangle.js:64]: `(0,0,0)` will cause troubles with `containsPoint()`. A degenerate triangle will contain any point.

Maybe that is why `(-2, -1, -1)` was selected -- so `containsPoint()` will return `false`.

Comment (WestLangley) [reply src/math/Triangle.js:64]: `(0,0,0)` will cause troubles with `containsPoint()`. (A degenerate triangle will contain any point.)

Maybe that is why `(-2, -1, -1)` was selected -- so `containsPoint()` will return `false`.

Comment (Mugen87) [reply src/math/Triangle.js:64]: Um, but @gkjohnson updated the method. It should return `false` because it checks for `null` now.
